core/trino-main/src/main/java/io/trino/execution/scheduler/StreamingStageExecution.java

```diff
@@ -0,0 +1,700 @@
+import io.trino.execution.RemoteTask;
+import io.trino.execution.SqlStageExecution;
+import io.trino.execution.StageId;
+import io.trino.execution.StateMachine;
+import io.trino.execution.StateMachine.StateChangeListener;
+import io.trino.execution.TaskId;
+import io.trino.execution.TaskState;
+import io.trino.execution.TaskStatus;
+import io.trino.execution.buffer.OutputBuffers;
+import io.trino.execution.buffer.OutputBuffers.OutputBufferId;
+import io.trino.failuredetector.FailureDetector;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.Split;
+import io.trino.spi.TrinoException;
+import io.trino.split.RemoteSplit;
+import io.trino.sql.planner.PlanFragment;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.RemoteSourceNode;
+import io.trino.util.Failures;
+import org.joda.time.DateTime;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.newConcurrentHashSet;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.ABORTED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.CANCELED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.FAILED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.FINISHED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.FLUSHING;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.PLANNED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.RUNNING;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.SCHEDULED;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.SCHEDULING;
+import static io.trino.execution.scheduler.StreamingStageExecution.State.SCHEDULING_SPLITS;
+import static io.trino.failuredetector.FailureDetector.State.GONE;
+import static io.trino.operator.ExchangeOperator.REMOTE_CONNECTOR_ID;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.spi.StandardErrorCode.REMOTE_HOST_GONE;
+import static java.util.Objects.requireNonNull;
+
+public class StreamingStageExecution
+        implements ResultsConsumer
```

Comment:
Instead of implementing this interface, maybe:

* Add a `getNotificationListener()` method
* Make whoever's creating these objects call getListener() and pass it to other instances that need to send notifications.